### PR TITLE
Notify followers when collaboration is accepted

### PR DESCRIPTION
**Pull request by Undeceiver**
Uses the same MapRelease notification type, but the message is slightly different, to account for the time that may pass between the map being published and the collaboration being accepted.

This is done in the same transaction that the collaboration is marked as accepted.

### CODE REVIEW

**Comment by Undeceiver** [comment]
image

File: src/jvmMain/kotlin/io/beatmaps/api/collaboration.kt [hunk 1]
```diff
@@ -81,17 +85,59 @@ fun Route.collaborationRoute() {
...
+                                it[accepted] = true
+                            },
+                            Collaboration.mapId
+                        )
+                    } else {
+                        Collaboration.deleteWhere {
+                            Collaboration.id eq req.collaborationId and (Collaboration.collaboratorId eq sess.userId)
+                        }
+
+                        null
```
Comment (Top-Cat) [inline src/jvmMain/kotlin/io/beatmaps/api/collaboration.kt:104]: Does this not mean that success is now always false for deletes (as mapId = null)

File: src/jvmMain/kotlin/io/beatmaps/api/collaboration.kt [hunk 2]
```diff
@@ -81,17 +85,59 @@ fun Route.collaborationRoute() {
...
+                        row[Follows.followerId].value
+                    }
+
+                    val nonRecipients = Follows.select {
+                        Follows.userId eq map.uploaderId and Follows.following
+                    }.map { row ->
+                        row[Follows.followerId].value
+                    } + listOf(map.uploaderId.value)
+
+                    val finalRecipients = recipients - nonRecipients.toSet()
```
Comment (Top-Cat) [inline src/jvmMain/kotlin/io/beatmaps/api/collaboration.kt:129]: Playing some more golf
It's somewhat complex but the database can quite efficiently do this filtering.

```kt
val followsAlias = Follows.alias("f2")
val recipients = Follows
    .join(followsAlias, JoinType.LEFT, followsAlias[Follows.followerId], Follows.followerId) {
        (followsAlias[Follows.userId] eq map.uploaderId) and followsAlias[Follows.following]
    }.select {
        followsAlias[Follows.id].isNull() and (Follows.followerId neq uploaderId) and
            (Follows.userId eq sess.userId) and Follows.upload and Follows.following
    }.map { row ->
        row[Follows.followerId].value
    }.toSet()
```

Comment (Top-Cat) [reply src/jvmMain/kotlin/io/beatmaps/api/collaboration.kt:129]: As per discussion on discord these should at least get `.slice()`'d

**Comment by Top-Cat** [comment]
Ok to test

**Pull request merged**